File: lexers/lex.py

rules = []
EOF = "\30"

def getNextStates(machStates, c):
    out = []
    for state in range(len(machStates)):
        if c in rules[state]['dfa']['sigma'] and machStates[state] != None:
            cindex = rules[state]['dfa']['sigma'].index(c)
            out.append(rules[state]['dfa']['deltaT'][machStates[state]][cindex])
        else:
            out.append(None)
    return out

def death(states) -> bool:
    return len(list(filter(lambda x: x != None, states))) == 0     

def getIndex(states) -> int:
    for i in range(len(states)):
        if states[i] != None:
            return i
    return -1

def start(input: str):
    lineNum = 1
    colNum = 1
    machStates = list(map(lambda x: x['dfa']['startState'], rules))
    token = ""
    bestFit = None
    tkstream = []

    while input != "":
        if input == "" or death(nextStates := getNextStates(machStates, input[0])):
            if bestFit == None:
                if token == "":
                    return []
                else:
                    raise Exception("Lex Error: Unexpected EOF in token \'" + token + "\'")
            else:
                if len(bestFit["symbol"]) >= 5 and bestFit["symbol"][:5] == "(ERR)":
                    print("Lex Error:" + bestFit["symbol"][5:])
                    return tkstream, False
                elif bestFit["symbol"] != "(SKIP)":
                    tkstream.append(bestFit)
                input = token[len(bestFit['lexeme']):] + input
                token = ""
                machStates = list(map(lambda x: x['dfa']['startState'], rules))
                bestFit = None
        else:
            machStates = nextStates
            token = token + input[0]
            if (i := getIndex(machStates)) != -1:
                bestFit = {"symbol":rules[i]["action"], "lexeme":token, "pos":[lineNum, colNum]}
            if input[0] == '\n': 
                lineNum += 1
                colNum = 1
            else:
                colNum += 1
            input = input[1:]
    else:
        if bestFit == None:
            raise Exception("Lex Error: Unexpected EOF in token \'" + token + "\'")
        tkstream.append(bestFit)

    
    tkstream.append({"symbol":EOF,"lexeme":None,"pos":[lineNum,colNum]})
    return tkstream

File: lexers/lex.go
package lex

type Automata struct {
	startState int
	finStates  []int
	action     string
	deltaT     map[int]map[rune]int
}

type Token struct {
	symbol string
	lexeme []rune
	pos    [2]int
}

func death(mach []int) bool {
	for _, v := range mach {
		if v != -1 {
			return false
		}
	}
	return true
}

func getIndex(spec []Automata, mach []int) int {
	for i := range spec {
		for _, f := range spec[i].finStates {
			if f == mach[i] {
				return i
			}
		}
	}
	return -1
}

func moveAhead(c rune, spec []Automata, mach []int) []int {
	out := make([]int, len(mach), cap(mach))
	for i, v := range mach {
		if v != -1 {
			if t, valid := spec[i].deltaT[v][c]; valid {
				out[i] = t
			} else {
				out[i] = -1
			}
		} else {
			out[i] = -1
		}
	}
	return out
}

func lex(spec []Automata, input []rune) (tkstream []Token, valid bool) {
	lineNum, colNum := 1, 1
	mach := make([]int, len(spec), len(spec))
	token := make([]rune, 0, cap(input))
	var bestFit *Token = nil

	for len(input) != 0 {
		if mv := moveAhead(input[0], spec, mach); len(input) == 0 || death(mv) {
			if bestFit == nil {
				if len(token) == 0 {
					return tkstream, true
				} else {
					return tkstream, false
				}
			} else {
				if len((*bestFit).symbol) >= 5 && (*bestFit).symbol[:5] == "(ERR)" {
					return tkstream, false
				} else if (*bestFit).symbol != "(SKIP)" {
					tkstream = append(tkstream, *bestFit)
				}
				input = append(token[len((*bestFit).lexeme):], input...)
				token = make([]rune, 0, cap(input))
				mach = make([]int, len(spec), len(spec))
				bestFit = nil
			}
		} else {
			mach = mv
			token = append(token, input[0])
			if i := getIndex(spec, mach); i != -1 {
				bestFit = &Token{spec[i].action, token, [2]int{lineNum, colNum}}
			}
			if input[0] == '\n' {
				lineNum += 1
				colNum = 1
			} else {
				colNum += 1
			}
			input = input[1:]
		}
	}
	if bestFit == nil {
		return tkstream, false
	}
	tkstream = append(tkstream, *bestFit)
	tkstream = append(tkstream, Token{"\x18", []rune{'\x18'}, [2]int{0, 0}})
	return tkstream, true
}
